tiledb/sm/query/query_condition.cc

```diff
@@ -669,49 +724,183 @@ QueryCondition::apply_clause(
+    const tdb_unique_ptr<ASTNode>& node,
+    const ArraySchema& array_schema,
+    uint64_t stride,
+    const std::vector<ResultCellSlab>& result_cell_slabs,
+    CombinationOp combination_op,
+    std::vector<uint8_t>& result_cell_bitmap) const {
+  if (!node->is_expr()) {
+    apply_ast_node(
+        node,
+        array_schema,
+        stride,
+        result_cell_slabs,
+        combination_op,
+        result_cell_bitmap);
+  } else {
+    switch (node->get_combination_op()) {
+        /*
+         * cl(q; a) means evaluate a clause (which may be compound) with query q
+         * given existing bitmap a
+         *
+         * Identities:
+         *
+         * cl(q; a) = cl(q; 1) /\ a
+         * cl1(q; a) /\ cl2(q; b) = cl1(q; cl2(q; a))
+         *
+         * cl1(q; a) \/ cl2(q; a) = (cl1(q; 1) /\ a) \/ (cl2(q; 1) /\ a)
+         *                        = (cl1(q; 1) \/ cl2(q; 1)) /\ a
+         */
+
+        /*
+         * cl1(q; a) /\ cl2(q; a) = cl1(q; cl2(q; a))
+         */
+      case QueryConditionCombinationOp::AND: {
+        if constexpr (std::
+                          is_same_v<CombinationOp, std::logical_and<uint8_t>>) {
+          for (const auto& child : node->get_children()) {
+            apply_tree(
+                child,
+                array_schema,
+                stride,
+                result_cell_slabs,
+                std::logical_and<uint8_t>(),
+                result_cell_bitmap);
+          }
+
+          // Handle the cl'(q, a) case
+        } else if constexpr (std::is_same_v<
+                                 CombinationOp,
+                                 std::logical_or<uint8_t>>) {
+          std::vector<uint8_t> combination_op_bitmap(
+              result_cell_bitmap.size(), 1);
+
+          for (const auto& child : node->get_children()) {
+            apply_tree(
+                child,
+                array_schema,
+                stride,
+                result_cell_slabs,
+                std::logical_and<uint8_t>(),
+                combination_op_bitmap);
+          }
+          for (size_t c = 0; c < combination_op_bitmap.size(); ++c) {
+            result_cell_bitmap[c] |= combination_op_bitmap[c];
+          }
+        }
+      } break;
+
+        /*
+         * cl1(q; a) \/ cl2(q; a) = a /\ (cl1(q; 1) \/ cl2(q; 1))
+         *                        = a /\ (cl1'(q; 0) \/ cl2'(q; 0))
+         *                        = a /\ (cl1'(q; cl2'(q; 0)))
+         */
+      case QueryConditionCombinationOp::OR: {
+        std::vector<uint8_t> combination_op_bitmap(
+            result_cell_bitmap.size(), 0);
+
+        for (const auto& child : node->get_children()) {
+          apply_tree(
+              child,
+              array_schema,
+              stride,
+              result_cell_slabs,
+              std::logical_or<uint8_t>(),
+              combination_op_bitmap);
+        }
+
+        /// TODO: We have not tested for optimization of & versus *.
```

Comment:
Could this be moved out of code and into a story?

Reply:
https://app.shortcut.com/tiledb-inc/story/18101/querycondition-test-optimization-of-versus yes